### ISSUE TITLE
isset() collision on OS X with protobuf 3.x

### ISSUE THREAD

**Issue by apjanke** [cruppstahl/upscaledb#82]
It looks like there is a collision problem with the `isset()` macro define in `src/0root/root.h` and the `<sys/params.h>` header on OS X, which is now included by protobuf version 3.0.

When I build upscaledb 2.2.0 on OS X 10.11.6 with Xcode 7.3.1 against protobuf 3.0.0 and libuv 0.10.37, the `samples/db1.c` test program crashes upon start, with what looks like it might be a segfault. We found this while working on the protobuf 2.x to 3.0 upgrade in Homebrew; details are at https://github.com/Homebrew/homebrew-core/pull/4145.

We tracked it down to a likely macro redefinition conflict. [Here](https://github.com/google/protobuf/blob/5d35e60900bf74e095dd98904d07f144dbd8d521/src/google/protobuf/stubs/port.h#L62), protobuf's port.h includes `<sys/params.h>`. And macOS's `<sys/param.h>` defines the following.

```
/* Bit map related macros. */
#define setbit(a,i)     (((char *)(a))[(i)/NBBY] |= 1<<((i)%NBBY))
#define clrbit(a,i)     (((char *)(a))[(i)/NBBY] &= ~(1<<((i)%NBBY)))
#define isset(a,i)      (((char *)(a))[(i)/NBBY] & (1<<((i)%NBBY)))
#define isclr(a,i)      ((((char *)(a))[(i)/NBBY] & (1<<((i)%NBBY))) == 0)
```

It looks like that redefines `isset` in an incompatible way.

We're currently trying to work around this by re-re-defining `isset` back to the upscaledb version, but it would be nice if this worked out of the box.

Might affect other BSD-based systems, too.

**Comment by ilovezfs** [comment]
>Might affect other BSD-based systems, too.

Yup: https://github.com/freebsd/freebsd/blob/master/sys/sys/param.h#L284-L285

**Comment by cruppstahl** [comment]
Thanks for the information. I'll replace the macro with a different name. Will provide a patch over the weekend.

**Comment by ilovezfs** [comment]
@cruppstahl thanks, that would be awesome! Bonus points if you tag a new release for us too :innocent:

**Comment by cruppstahl** [comment]
I've updated the master branch and topic/2.2.0 (they're identical). Is this
enough or do you need anything else? I would like to wait with releasing
2.2.1 (the next version) because it has a number of patches that were not
properly tested...

2016-09-09 8:47 GMT+02:00 ilovezfs <notifications@github.com>:

**Comment by apjanke** [comment]
Nope. As long as we know the patch is getting merged, we can wait until the next point release comes out normally. We already have a workaround in place on Homebrew's side for the time being.

Thanks!
